crates/orchestrator/src/network_spec/node.rs

```diff
@@ -0,0 +1,223 @@
+use configuration::shared::{
+    node::{EnvVar, NodeConfig},
+    resources::Resources,
+    types::{Arg, AssetLocation, Command, Image},
+};
+use multiaddr::Multiaddr;
+
+use crate::{
+    errors::OrchestratorError,
+    generators,
+    network::AddNodeOpts,
+    shared::types::{ChainDefaultContext, NodeAccounts, ParkedPort},
+};
+
+/// A node configuration, with fine-grained configuration options.
+#[derive(Debug, Clone)]
+pub struct NodeSpec {
+    /// Node name (should be unique or an index will be appended).
+    pub(crate) name: String,
+
+    /// Node key, used for compute the p2p identity.
+    pub(crate) key: String,
+
+    // libp2p local identity
+    pub(crate) peer_id: String,
+
+    /// Accounts to be injected in the keystore.
+    pub(crate) accounts: NodeAccounts,
+
+    /// Image to run (only podman/k8s). Override the default.
+    pub(crate) image: Option<Image>,
+
+    /// Command to run the node. Override the default.
+    pub(crate) command: Command,
+
+    /// Arguments to use for node. Appended to default.
+    pub(crate) args: Vec<Arg>,
+
+    /// Wether the node is a validator.
+    pub(crate) is_validator: bool,
+
+    /// Whether the node keys must be added to invulnerables.
+    pub(crate) is_invulnerable: bool,
+
+    /// Whether the node is a bootnode.
+    pub(crate) is_bootnode: bool,
+
+    /// Node initial balance present in genesis.
+    pub(crate) initial_balance: u128,
+
+    /// Environment variables to set (inside pod for podman/k8s, inside shell for native).
+    pub(crate) env: Vec<EnvVar>,
+
+    /// List of node's bootnodes addresses to use. Appended to default.
+    pub(crate) bootnodes_addresses: Vec<Multiaddr>,
+
+    /// Default resources. Override the default.
+    pub(crate) resources: Option<Resources>,
+
+    /// Websocket port to use.
+    pub(crate) ws_port: ParkedPort,
+
+    /// RPC port to use.
+    pub(crate) rpc_port: ParkedPort,
+
+    /// Prometheus port to use.
+    pub(crate) prometheus_port: ParkedPort,
+
+    /// P2P port to use.
+    pub(crate) p2p_port: ParkedPort,
+
+    /// libp2p cert hash to use with `webrtc` transport.
+    pub(crate) p2p_cert_hash: Option<String>,
+
+    /// Database snapshot. Override the default.
+    pub(crate) db_snapshot: Option<AssetLocation>,
+}
+
+impl NodeSpec {
+    pub fn from_config(
+        node_config: &NodeConfig,
+        chain_context: &ChainDefaultContext,
+    ) -> Result<Self, OrchestratorError> {
+        // Check first if the image is set at node level, then try with the default
+        let image = if let Some(img) = node_config.image() {
+            Some(img.clone())
+        } else {
+            chain_context.default_image.cloned()
+        };
```

Comment:
nit: Can be simplified


```suggestion
        let image = node_config.image().or(chain_context.default_image).cloned();
```